libpod/container_internal_linux.go

```diff
@@ -744,7 +744,12 @@ func (c *Container) generateSpec(ctx context.Context) (*spec.Spec, error) {
 
 	// Warning: CDI may alter g.Config in place.
 	if len(c.config.CDIDevices) > 0 {
-		if err = cdi.UpdateOCISpecForDevices(g.Config, c.config.CDIDevices); err != nil {
+		registry := cdi.GetRegistry()
+		unresolved, err := registry.InjectDevices(g.Config, c.config.CDIDevices...)
+		if len(unresolved) > 0 {
+			logrus.Debugf("Could not resolve CDI devices: %v", unresolved)
```

Comment:
Yes, this should probably be a warning. This would indicate that the state of the registry has changed between determining the list of CDI devices (where `GetDevice` is called) and here where the actual injection takes place.

One question I did have was whether there would be a recommended way to propagate a reference to the registry from `MakeContainer` in [`create_container.go`](https://github.com/containers/podman/pull/12825/files#diff-3ff238a49e99bd584fa865688b0fd9914acd5cdd71f46606c6750a855da6cf4fR169) to the use here in `generateSpec`?

Reply:
The container's image name and image ID are stored in its configuration at this point, but there's no guarantee those are actually populated, Podman supports creating images from user-specified root filesystems.

Reply:
By "registry" in this case I mean a cache of the CDI devices defined by spec files available on the system (see the interface [here](https://github.com/container-orchestrated-devices/container-device-interface/blob/master/pkg/cdi/registry.go#L33). The idea is that spec files in, for example, `/etc/cdi/*.json` are loaded once and then these specifications are used.

Reply:
Ahh. Sorry, registry usually means a very different thing for containers.

I checked, and there's no location for the spec files currently listed that I can find. If one were to be added, `containers.conf` would be the most logical location - that's a centralized config file that we could put the path in, that all containers can access

Reply:
I think I'm still miscommunicating what I need to know.

When processing the original list of devices from the config in `create_container.go`, we instantiate a cache that stores the state of the CDI specs for the system (defining where these are read from is currently out of scope). The question is whether there is a way to pass this **instance** of the cache along so that this is also used in `generateSpec` in `container_internal_linux.go`? The motivation for this is that the files defining the CDI definition may have changed on disk so that the list of CDI devices being processed here does not have the desired effect -- hence the warning.

To be clear, I don't think this is a blocker for this particular PR, but it may be good to consider how to plumb this through in future if there is something that makes sense and does not add too much additional overhead. On the other hand, maybe @klihub has insights into the lifetime of the singleton instantiated in the `cdi` package when used at these two locations -- if we have some guarantees that these are in fact the same object then there would be no need to pass these through explicitly (except making the link clearer).

Reply:
There are temporary files directories that you could use, but they're all internal to Libpod, so accessing them from `pkg/specgen` could prove to be a problem. Is this a global directory, or a per-container one? We could definitely do a global CDI state directory inside Libpod's temporary files directory, and expose that to the world.

Reply:
This is a global directory. For what it's worth, I have changed the implementation to only look at whether the device specified is a valid CDI device name. This means that we are no longer instantiating this cache twice. I think this is sufficient for the time being. If this becomes a requirement we can open a discussion as to how this should be done correctly.